CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+
+This serves the exact purpose that `PGame` does for `IGame`. However, unlike `PGame`'s relatively
+short construction, we must prove some extra definitions for computing on top of `SGame`.
+
+This could perfectly well have been in `Type 0`, but we make it universe polymorphic for
+convenience when building quotient types. However, conversions from computable games to their
+noncomputable counterparts are squeezed to `Type 0`. -/
+inductive SGame : Type u
+  | mk (m n : ℕ) (f : Fin m → SGame) (g : Fin n → SGame) : SGame
+compile_inductive% SGame
+
+namespace SGame
+
+/-- The number of left moves on a `SGame`. -/
+def LeftMoves : SGame → ℕ
+  | mk m _ _ _ => m
+
+/-- The number of right moves on a `SGame`. -/
+def RightMoves : SGame → ℕ
+  | mk _ n _ _ => n
+
+/-- Perform a left move. -/
+def moveLeft : ∀ g : SGame, Fin g.LeftMoves → SGame
+  | mk _ _ f _ => f
+
+/-- Perform a right move. -/
+def moveRight : ∀ g : SGame, Fin g.RightMoves → SGame
+  | mk _ _ _ g => g
+
+@[simp] theorem leftMoves_mk (m n f g) : (mk m n f g).LeftMoves = m := rfl
+@[simp] theorem rightMoves_mk (m n f g) : (mk m n f g).RightMoves = n := rfl
+@[simp] theorem moveLeft_mk (m n f g) : (mk m n f g).moveLeft = f := rfl
+@[simp] theorem moveRight_mk (m n f g) : (mk m n f g).moveRight = g := rfl
+
+/-- A well-founded relation on `SGame`.
+While this doesn't have an equivalent in `PGame`, this is necessary to do
+nice well-defined recursion in `SGame`. -/
+inductive IsOption : SGame → SGame → Prop
+  | moveLeft {x : SGame} (n : Fin x.LeftMoves) : IsOption (x.moveLeft n) x
+  | moveRight {x : SGame} (n : Fin x.RightMoves) : IsOption (x.moveRight n) x
+
+theorem isOption_wf : WellFounded IsOption := by
+  refine ⟨rec fun s t f g IHl IHr ↦ .intro _ ?_⟩
+  rintro y (h | h)
+  · exact IHl _
+  · exact IHr _
+
+instance : WellFoundedRelation SGame := ⟨_, isOption_wf⟩
+
+macro "sgame_wf" : tactic =>
+  `(tactic| all_goals solve_by_elim
+    [Prod.Lex.left, Prod.Lex.right, PSigma.Lex.left, PSigma.Lex.right,
+    IsOption.moveLeft, IsOption.moveRight] )
+
+instance instDecidableEq : ∀ x y : SGame, Decidable (x = y)
+  | mk m n f g, mk m' n' f' g' => if h : m = m' ∧ n = n' then by
+    have : ∀ a b, Decidable (f a = f' b) := fun a b ↦ instDecidableEq ..
+    have : ∀ a b, Decidable (g a = g' b) := fun a b ↦ instDecidableEq ..
+    rw [mk.injEq, Fin.heq_fun_iff h.1, Fin.heq_fun_iff h.2]
+    infer_instance
+  else .isFalse (by simp_all)
+
+/-- The identical relation on short games. See `SGame.Identical`. -/
+def Identical : SGame.{u} → SGame.{u} → Prop
+  | mk _ _ xL xR, mk _ _ yL yR =>
+      Relator.BiTotal (fun i j ↦ Identical (xL i) (yL j)) ∧
+      Relator.BiTotal (fun i j ↦ Identical (xR i) (yR j))
+
+@[inherit_doc] scoped infix:50 " ≡ " => SGame.Identical
+
+theorem identical_iff : ∀ {x y : SGame}, x ≡ y ↔
+    Relator.BiTotal (x.moveLeft · ≡ y.moveLeft ·) ∧ Relator.BiTotal (x.moveRight · ≡ y.moveRight ·)
+  | mk .., mk .. => Iff.rfl
```

Comment:
```suggestion
  | mk .., mk .. => .rfl
```

Reply:
I'll also change that here:

https://github.com/vihdzp/combinatorial-games/blob/627b3ecb25bb4ce1bff9a5e6445463c229603bc8/CombinatorialGames/Game/IGame.lean#L143